File: B1.ts
/**
 * @author Marco Cantusci
 * @version 1.0.0
 * @date 2025-12-10
 * @fileoverview This program keeps track of car stats.
 */

// constants and variables
let odometer: number = 65000.0; // mileage of Car
let oilChangeKM: number = 65000.0; // value since the last oil change
let carColor: string = "Black"; // color of Car
const carModel: string = "Honda Civic"; // model of Car
let newMileage: number = 0.0; // new mileage amount
const gasCost: number[] = new Array(10); // cost of gas per fill up.
let gasIndex: number = 0;

// carsStats function(2)
function carStats(carModel: string, carColor: string, odometer: number, oilChangeKM: number): string {
  let stats = "";
  stats += `\nCar Model: ${carModel}`;
  stats += `\nCar Colour: ${carColor}`;
  stats += `\nOdometer: ${odometer}`;
  stats += `\nLast Oil Change: ${oilChangeKM} km`;
  return stats;
}

// wrapCar function(3)
function wrapCar(): string {
  const newColor = prompt("Enter a new color to wrap your car: ") ||
    ("\n");
  return newColor;
}

// drive function(4)
function drive(odometer: number, newMileage: number) {
  const randomKM = Math.floor(Math.random() * (1000 - 100 + 1) + 100);
  newMileage = odometer + randomKM;
  return randomKM;
}

// fillUp function(5)
function fillUp(gasCost: number[], gasIndex: number) {
  let loop: string = "yes";
  while (loop.toLowerCase() == "yes") {
    let priceString: string =
      prompt(`Cost to fill up gas ${gasIndex + 1}: $`) || "0";

    while (parseFloat(priceString) <= -1) {
      priceString = prompt("Invalid input! Please enter a positive number: ") ||
        "0";
    }

    const priceNumber: number = parseFloat(priceString);
    gasCost[gasIndex] = priceNumber;
    gasIndex++;

    loop = prompt("Add another fill up? (yes/no): ") || ("no");
    loop.toLowerCase();

    while (loop === null || loop !== "yes" && loop !== "no") {
      loop = prompt("Invalid input. Please enter 'yes' or 'no':") || ("no");
    }
  }

  return gasIndex;
}

// displayCostToFillUp function(6)
function displayCostToFillUp(gasCost: number[], gasIndex: number): number {
  let counting: number = 0;
  let total: number = 0;

  for (let counter = 0; counter < gasIndex; counter++) {
    console.log("\n");
    console.log(`Fill up ${counter + 1}: $${gasCost[counter]}`);
    total += gasCost[counter];
    counting++;
  }

  const average: number = total / counting;
  return average;
}

// oilChange function(7)
function oilChange(mileage: number, oilChangeKM: number): boolean {
  if ((mileage - oilChangeKM) >= 5000) {
    console.log("An oil change was done.");
    return true;
  } else {
    return false;
  }
}

// output
// print carStats
console.log("Initial Car Stats");
console.log(carStats(carModel, carColor, odometer, oilChangeKM));

// print drive function
const driven: number = drive(odometer, newMileage);
console.log(`You drove ${driven} km`);
odometer += driven;

// print oilChange function
if (oilChange(odometer, oilChangeKM)) {
  oilChangeKM = odometer;
} else {
  console.log("Your car does not need an oil change.");
}

// print wrapCar function
carColor = wrapCar();

// print fillUp function
gasIndex = fillUp(gasCost, gasIndex);

// print gas costs/average
console.log(`\nAverage cost to fill up gas: $${displayCostToFillUp(gasCost, gasIndex)}`);

// print updated car stats
console.log("\n", carStats(carModel, carColor, odometer, oilChangeKM));

console.log("\nDone.");

File: B1.go
/**
 * @author Marco Cantusci
 * @version 1.0.0
 * @date 2025-12-10
 * @fileoverview This program keeps track of car stats.
 */

package main

import (
	"bufio"
	"fmt"
	"math"
	"math/rand/v2"
	"os"
	"strconv"
	"strings"
)

// constants and variables
var odometer float64 = 65000.0        // mileage of Car
var oilChangeKM float64 = 65000.0     // value since the last oil change
var carColor string = "Black"         // color of Car
const carModel string = "Honda Civic" // model of Car
var gasCost [10]float64               // cost of gas per fill up.
var gasIndex int = 0

var reader = bufio.NewReader(os.Stdin)

// carsStats function(2)
func carStats(carModel string, carColor string, odometer float64, oilChangeKM float64) string {
	var stats string = ""
	stats += fmt.Sprintf("\nCar Model: %s", carModel)
	stats += fmt.Sprintf("\nCar Colour: %s", carColor)
	stats += fmt.Sprintf("\nOdometer: %.2f", odometer)
	stats += fmt.Sprintf("\nLast Oil Change: %.2f km", oilChangeKM)
	return stats
}

// wrapCar function(3)
func wrapCar() string {
	fmt.Print("Enter a new colour to wrap your car: ")
	color, _ := reader.ReadString('\n')
	color = strings.TrimSpace(color)

	if color == "" {
		return carColor
	}
	return color
}

// drive function(4)
func drive() float64 {
	var randomKM float64 = rand.Float64()
	randomKM = math.Floor(randomKM*901) + 100
	return randomKM
}

// fillUp function(5)
func fillUp(gasCost [10]float64, gasIndex int) ([10]float64, int) {
	var loop string = "yes"

	for strings.ToLower(loop) == "yes" {
		fmt.Printf("Cost to fill up gas %d: $", gasIndex+1)
		priceString, _ := reader.ReadString('\n')
		priceString = strings.TrimSpace(priceString)
		priceNumber, _ := strconv.ParseFloat(priceString, 64)

		for priceNumber < 0 {
			fmt.Print("Invalid input! PLease enter a positive number: ")
			priceString, _ = reader.ReadString('\n')
			priceString = strings.TrimSpace(priceString)
			priceNumber, _ = strconv.ParseFloat(priceString, 64)
		}

		gasCost[gasIndex] = priceNumber
		gasIndex++

		fmt.Print("Add another fill up? (yes/no): ")
		loop, _ = reader.ReadString('\n')
		loop = strings.TrimSpace(loop)

		for loop != "yes" && loop != "no" {
			fmt.Print("Invalid input! Please enter 'yes' or 'no': ")
			loop, _ = reader.ReadString('\n')
			loop = strings.TrimSpace(loop)
		}
	}
	return gasCost, gasIndex
}

// displayCostToFillUp function(6)
func displayCostToFillUp(gasCost [10]float64, gasIndex int) float64 {
	var total float64 = 0.0

	for counter := 0; counter < gasIndex; counter++ {
		fmt.Println()
		fmt.Printf("Fill up %d: $%.2f\n", counter+1, gasCost[counter])
		total += gasCost[counter]
	}

	if gasIndex == 0 {
		return 0
	}

	var average float64 = total / float64(gasIndex)
	return average
}

// oilChange function(7)
func oilChange(mileage float64, lastOilChange float64) bool {
	if (mileage - lastOilChange) >= 5000 {
		fmt.Println("An oil change was done.")
		return true
	}
	return false
}

// output
func main() {

	// print carStats
	fmt.Println("Initial Car Stats")
	fmt.Println(carStats(carModel, carColor, odometer, oilChangeKM))

	// print drive function
	var driven float64 = drive()
	fmt.Printf("You drove %.0f km\n", driven)
	odometer += driven

	// print oilChange function
	if oilChange(odometer, oilChangeKM) {
		oilChangeKM = odometer
	} else {
		fmt.Println("Your car does not need an oil change.")
	}

	// print wrapCar function
	carColor = wrapCar()

	// print fillUp function
	gasCost, gasIndex = fillUp(gasCost, gasIndex)

	// print gas costs/average
	fmt.Printf("\nAverage cost to fill up gas: $%.2f\n", displayCostToFillUp(gasCost, gasIndex))

	// print updated car stats
	fmt.Println("\nUpdated Car Stats")
	fmt.Println(carStats(carModel, carColor, odometer, oilChangeKM))

	fmt.Println("\nDone.")
}
